### ISSUE TITLE
SS_doReto: Add option to adjust the late recruitment start year

### ISSUE THREAD

**Issue by CWetzel** [r4ss/r4ss#245]
It would be nice if there was an option to change the late recruitment estimation year to change according to the retrospective year.

**Comment by taylori** [comment]
Good idea. This is something that we probably all should be doing but aren't. Another element in the mix is the bias adjustment parameters, where the quantities `last_yr_fullbias_adj_in_MPD` and `first_recent_yr_nobias_adj_in_MPD` should typically also be shifted back in parallel with the late recruitment start year. 

This is something that might be worth building into SS rather than relying on r4ss, so we should discuss with Rick. The r4ss functions for reading and writing control files for SS version 3.30 models aren't necessarily robust enough to rely on them to make these changes.

**Comment by iantaylor-NOAA** [comment]
Email discussion with @Cole-Monnahan-NOAA, @Rick-Methot-NOAA, and @kellijohnson-NOAA suggests that there's a critical mass of interest in moving this topic forward. I've renamed to issue to clarify that there are multiple model settings which might need to be adjusted in a retrospective. It's not clear which of these are likely to make much difference, but automated tools to adjust them if requested would make it easier to answer that question.

Settings include
- main/late recruit start year
- recdev bias adjustment
- time blocks and other time-varying parameter settings

**Comment by IJStewart** [comment]
Automation of the structural components (time-varying parameters) that must be adjusted will be a welcome improvement. 
Recruitment vectors and bias corrections should probably have a caution flag on them though - it is common for assessment data to lag a year in things like ages for some or all fleets, which could suggest the need for a more complex approach than just removing 1 year from the main recruitment deviation and/or bias correction settings when doing retrospectives.

**Comment by akatan999** [comment]
I did a lpt of testing on this issue. In general, it does not matter when doing retro, but sometime (thua quite rarely) it does. This happens in particularly when there is a very large or very small year class in the retrospective period, when the estimated bias adjustment is close to 1 and when using time-varying parameter as for example random walk in key parameters as selectivity or growth. Indeed, having the possibility to test it in an utomatic way would be a great improvement.

**Comment by paulburch** [comment]
Hi there,

I've had a go at fixing the issue with estimating to many rec devs, it's a bit of a hack and depends on editing the ctl file, but I noticed you've done something similar for time blocking so you may want to use it as a temporary solution.

I've created a modified version of the SS_doRetro function (attached) that removes the number years of retrospective years from the last year to estimate rec devs in the ctl file. I added the two lines below to the code that edits the ctl file so it now removes the # retro years from the year of the last rec dev (note it greps for the text “last year of main recr_devs” so that needs to be in the ctl file).
 
    rec_dev_str <- ctl[grep("last year of main recr_devs", ctl)]
    ctl[grep("last year of main recr_devs", ctl)] <- paste0(as.numeric(substr(rec_dev_str, 1,4)) + years[iyr], " # last year of main recr_devs; forecast devs start in following year ")
 
The SESSF Team down in Hobart has tested it on three of our stocks and it seems to work fine. Andre suggested adding an argument to the function to ensure backwards compatibility which I'm happy to add if you want to include my suggested fix.

Cheers,
Paul

[SS_doRetro_alt.txt](https://github.com/r4ss/r4ss/files/7465573/SS_doRetro_alt.txt)
n

**Comment by akatan999** [comment]
It is indeed a very good idea to have the possiblity to control the number of recdev in retrospective analysis instead of doing it manually. Thus, adding it to SS_doRetro would be ideal. Thanks Paul for the fucntion, I will test it with some of my problematic retro stocks.

Cheers
Max

**Comment by paulburch** [comment]
You could probably do the same thing to the last year if the bias ramp

**Comment by iantaylor-NOAA** [comment]
Hi @paulburch,
Thanks for contributing the modified function. Your changes to the main recdevs look good. Before merging into r4ss, I might try to add the change to the bias ramp and also add a new argument to give user's control of whether this happens or not. I don't know if this would be too busy, but it might be appropriate to include multiple new arguments, such as `shift_main_recdevs = TRUE/FALSE`, `shift_bias_adjust = TRUE/FALSE`, and `first_retro_with_shift = -1`, where that last argument (which could get a shorter name) would cover the case that @IJStewart notes where you might only be losing data and wanting to shift the recdevs and bias adjustment when you get to retro year = -2.

@kellijohnson-NOAA, I know you did some work trying to deal with time-varying parameters and blocks in retrospectives. Please comment if you're interested in trying to bring your changes into r4ss (presumably after lingcod is finally over).

**Comment by paulburch** [comment]
Hi @iantaylor-NOAA,

Those suggestions look great. I can have a go at implementing them in a week or so if you want?

**Comment by iantaylor-NOAA** [comment]
Flagging here that a fix to this issue is proposed in Pull Request #1087. Comment there if the implementation doesn't look right or work right for you.
